tests/Unit/API/Trace/TraceStateTest.php

```diff
@@ -210,4 +213,58 @@ public function test_validate_value(): void
         $this->assertSame(257, strlen($invalidValue));
         $this->assertNull($tracestate->get($validKey));
     }
+
+    #[DataProvider('toStringProvider')]
+    public function test_to_string(array $entries, ?int $limit, string $expected): void
+    {
+        $traceState = new TraceState();
+        foreach (array_reverse($entries) as $key => $value) {
+            $traceState = $traceState->with($key, $value);
+        }
+
+        $this->assertSame($expected, $traceState->toString(limit: $limit));
+    }
+
+    public static function toStringProvider(): array
+    {
+        return [
+            [[], null, ''],
+            [['key' => 'value'], null, 'key=value'],
+            [['key' => 'value', 'key2' => 'value2'], null, 'key=value,key2=value2'],
+            [['key' => 'value', 'key2' => 'value2'], 21, 'key=value,key2=value2'],
+            [['key' => 'value', 'key2' => 'value2'], 14, 'key=value'],
+            [['key' => 'value', 'key2' => 'value2'], 9, 'key=value'],
+            [['key' => 'value', 'key2' => 'value2'], 5, ''],
+            [['key' => 'value', 'a' => 'b'], 5, ''],
+            [[str_repeat('a', 10) => str_repeat('v', 50), 'key' => 'value', 'key2' => 'value2'], 50, ''],
+            [[str_repeat('a', 10) => str_repeat('v', 150), 'key' => 'value', 'key2' => 'value2'], 50, 'key=value,key2=value2'],
+            [[str_repeat('a', 10) => str_repeat('v', 117), 'key' => 'value', 'key2' => 'value2'], 50, ''],
+            [[str_repeat('a', 10) => str_repeat('v', 118), 'key' => 'value', 'key2' => 'value2'], 50, 'key=value,key2=value2'],
+            [[str_repeat('a', 10) => str_repeat('v', 118), 'key' => 'value', 'key2' => 'value2'], 14, 'key=value'],
+        ];
+    }
+
+    #[DataProvider('parseProvider')]
+    public function test_parse(string $tracestate, ?string $expected): void
+    {
+        $traceState = new TraceState($tracestate);
+        $this->assertSame($expected ?? '', (string) $traceState);
+    }
+
+    public static function parseProvider(): array
+    {
+        return [
+            ['key=value', 'key=value'],
+            ['key=value,key2=value2', 'key=value,key2=value2'],
+            ['key=value,key2=value2,key=value3', 'key=value,key2=value2'],
+
+            ['', ''],
+            ['  ', ''],
+            ['key=value,,key2=value2', 'key=value,key2=value2'],
+            ['key=value,   ,key2=value2', 'key=value,key2=value2'],
+
+            ['0', null],
```

Comment:
Doesn't seem like there is any reason to differentiate between `''` and `null` here, can change to `''` and remove nullability from test method.

Reply:
Will add an assertion that a warning is triggered if expected is `null`.